Rebuild Components correctly for collapsed Networks

tcstewar opened the pull request:
Fixes #586

Seanny123 commented on nengo_gui/components/netgraph.py:
```diff
@@ -162,9 +162,41 @@ def _reload(self, code=None):
+                    try:
+                        old_obj = eval(item, old_locals)
+                    except:
+                        continue
```
Why would evaluating this cause an exception? If I understood why, I would feel more comfortable with behaviour of continuing regardless of exceptions.

Also, if you do continue after an exception, isn't that going to cause an error in line Line 188 where you do `not isinstance(new_obj, old_obj.__class__)` since `old_obj` won't exist?

tcstewar replied:
> Also, if you do continue after an exception, isn't that going to cause an error in line Line 188 where you do not isinstance(new_obj, old_obj.__class__) since old_obj won't exist?

That's not what ```continue``` does.  You're thinking of ```pass```, which would cause the problem you're indicating.  ```continue``` causes python to go on to the next item in the ```for``` loop.

> Why would evaluating this cause an exception? If I understood why, I would feel more comfortable with behaviour of continuing regardless of exceptions.

I should definitely add a comment about this.  Basically, the list of things returned by ```c.code_python_args``` can contain things that aren't items.  It includes all the arguments that should be passed into the Component, including things like ```target='default'``` for the semantic pointer plots, for example.  Evaluating those will cause an Exception (since it's not a valid expression), so we should skip those.

That said, I should probably change this to be a bit more robust for different cases of things that might be in ```c.code_python_args```, as now that you've pointed it out I can think of a few cases that would break this logic (but they're all cases that we don't use at the moment).

tcstewar commented:
I've reworked the Component argument checking code to handle the generic case, so it no longer needs that ```continue``` hack (which was basically it giving up on anything that was too hard to parse).  

I also fixed a bug where it was using the locals from the immediately previous run of the script for comparison, rather than the previous *successful* run of the script (which is the last time the netgraph got updated, since it only gets updated when there's no errors).  That might have been causing all sorts of other weird bugs.....

Seanny123 commented on nengo_gui/components/netgraph.py:
```diff
@@ -162,9 +162,52 @@ def _reload(self, code=None):
+                        new_obj = None
+
+                    keep_object = True
+                    if new_obj is None:
```
Since this code is identical to the code in line 128, I'd like to throw the whole `keep_object` decision tree into a function.

tcstewar replied:
It's not identical.  The code around 128 doesn't need the distinction between removed_items and rebuilt_items, so it doesn't do that.  It'd certainly be possible and a good idea to refactor this stuff in general, such that it'd be possible to have identical code here, but I think that's outside the scope of this bugfix.

That said, the keep_object=True line should be removed...  :)

Seanny123 replied:
Sounds good to me.

Seanny123 commented:
LGTM.
